dnsprovider/pkg/dnsprovider/providers/scaleway/dns.go

```diff
@@ -0,0 +1,571 @@
+	}
+
+	if len(r.upserts) > 0 {
+		for _, rrset := range r.upserts {
+			err := r.applyResourceRecordSet(rrset)
+			if err != nil {
+				return fmt.Errorf("failed to apply resource record set: %s, err: %s", rrset.Name(), err)
+			}
+		}
+
+		klog.V(2).Info(ctx, "record change set upserts complete")
+	}
+
+	if len(r.removals) > 0 {
+		records, err := getRecords(r.client, r.zone.Name())
+		if err != nil {
+			return err
+		}
+
+		for _, record := range r.removals {
+			for _, domainRecord := range records {
+				if domainRecord.Name == record.Name() {
+					err := deleteRecord(r.client, r.zone.Name(), domainRecord.ID)
+					if err != nil {
+						return fmt.Errorf("failed to delete record: %v", err)
+					}
+				}
+			}
+		}
+
+		klog.V(2).Info(ctx, "record change set removals complete")
+	}
+
+	klog.V(2).Info(ctx, "record change sets successfully applied")
+	return nil
+}
+
+// IsEmpty returns true if a changeset is empty, false otherwise
+func (r *resourceRecordChangeset) IsEmpty() bool {
+	if len(r.additions) == 0 && len(r.removals) == 0 && len(r.upserts) == 0 {
+		return true
+	}
+
+	return false
+}
+
+// ResourceRecordSets returns the associated resourceRecordSets of a changeset
+func (r *resourceRecordChangeset) ResourceRecordSets() dnsprovider.ResourceRecordSets {
+	return r.rrsets
+}
+
+// applyResourceRecordSet will create records of a domain as required by resourceRecordChangeset
+// and delete any previously created records matching the same name.
+// This is required for scaleway since it's API does not handle record sets, but
+// only individual records
+func (r *resourceRecordChangeset) applyResourceRecordSet(rrset dnsprovider.ResourceRecordSet) error {
+	deleteRecords, err := getRecordsByName(r.client, r.zone.Name(), rrset.Name())
+	if err != nil {
+		return fmt.Errorf("failed to get record IDs to delete")
+	}
+
+	addRecords := []*domain.Record(nil)
+
+	for range rrset.Rrdatas() {
+		for _, rrdata := range rrset.Rrdatas() {
+			addRecords = append(addRecords, &domain.Record{
+				Name: rrset.Name(),
+				Data: rrdata,
+				TTL:  uint32(rrset.Ttl()),
+				Type: domain.RecordType(rrset.Type()),
+			})
+		}
+	}
+
+	recordCreateRequest := &domain.UpdateDNSZoneRecordsRequest{
+		DNSZone: r.zone.parentDomain,
+		Changes: []*domain.RecordChange{
+			{
+				Add: &domain.RecordChangeAdd{
+					Records: addRecords,
+				},
+			},
+		},
+	}
+
+	_, err = createRecord(r.client, recordCreateRequest)
+	if err != nil {
+		return fmt.Errorf("could not create record: %v", err)
+	}
+
+	for _, record := range deleteRecords {
+		err = deleteRecord(r.client, r.zone.Name(), record.ID)
+		if err != nil {
+			return fmt.Errorf("error cleaning up old records: %v", err)
+		}
+	}
+
+	return nil
+}
+
+// listDomains returns a list of scaleway Domain objects
+//func listDomains(c *scw.Client) ([]*domain.DNSZone, error) {
+func listDomains(c *scw.Client) ([]*domain.DomainSummary, error) {
+	registrarApi := domain.NewRegistrarAPI(c)
```

Comment:
Why do we need the registrar API here? We are not using this one to pilot a domain name we already own.